### PR TITLE
HtoE and Normalization of M field

### PR DESCRIPTION
**Pull request by ruohai0925**
Add H_maxwell = B/mu0 - M in the MacroscopicEvolveE.cpp;

Add M normalization in the MacroscopicEvolveM.cpp;

Some other small modifications;

//
This PR has been checked before. See here: (https://github.com/jackieyao0114/WarpX/pull/4). It can be merged very quickly.

### CODE REVIEW

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp [hunk 1]
```diff
@@ -43,6 +43,7 @@ void FiniteDifferenceSolver::MacroscopicEvolveM (
         amrex::Real const dt,
         std::unique_ptr<MacroscopicProperties> const& macroscopic_properties )
     {
+    amrex::Real mag_normalized_error = 0.1; // normalization error of M field for checking
```
Comment (jackieyao0114) [inline Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp]: Can we make ``mag_normalized_error`` a user-input?

Comment (ajnonaka) [reply Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp]: I'm fine if we make this a very short, separate PR as well.

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp [hunk 2]
```diff
@@ -125,6 +126,22 @@ void FiniteDifferenceSolver::MacroscopicEvolveM (
                 + dt * Gil_damp * ( M_xface(i, j, k, 0) * ( M_xface(i, j, k, 2) * Hx_eff - M_xface(i, j, k, 0) * Hz_eff)
                 - M_xface(i, j, k, 1) * ( M_xface(i, j, k, 1) * Hz_eff - M_xface(i, j, k, 2) * Hy_eff));
 
+
+              // temporary normalized magnitude of M_xface field at the fixed point
+              // re-investigate the way we do Ms interp, in case we encounter the case where Ms changes across two adjacent cells that you are doing interp
+          amrex::Real mag_normalized = std::sqrt( std::pow(M_xface(i, j, k, 0),2.0) + std::pow(M_xface(i, j, k, 1),2.0) +
```
Comment (ajnonaka) [inline Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp]: fix the indentation

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp [hunk 3]
```diff
@@ -170,6 +187,22 @@ void FiniteDifferenceSolver::MacroscopicEvolveM (
                 + dt * Gil_damp * ( M_yface(i, j, k, 0) * ( M_yface(i, j, k, 2) * Hx_eff - M_yface(i, j, k, 0) * Hz_eff)
                 - M_yface(i, j, k, 1) * ( M_yface(i, j, k, 1) * Hz_eff - M_yface(i, j, k, 2) * Hy_eff));
 
+
+              // temporary normalized magnitude of M_yface field at the fixed point
+              // re-investigate the way we do Ms interp, in case we encounter the case where Ms changes across two adjacent cells that you are doing interp
+          amrex::Real mag_normalized = std::sqrt( std::pow(M_yface(i, j, k, 0),2.0) + std::pow(M_yface(i, j, k, 1),2.0) +
```
Comment (ajnonaka) [inline Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp]: indentation

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp [hunk 4]
```diff
@@ -214,7 +247,22 @@ void FiniteDifferenceSolver::MacroscopicEvolveM (
               M_zface(i, j, k, 2) += dt * (PhysConst::mu0 * mag_gamma_interp) * ( M_zface(i, j, k, 0) * Hy_eff - M_zface(i, j, k, 1) * Hx_eff)
                 + dt * Gil_damp * ( M_zface(i, j, k, 0) * ( M_zface(i, j, k, 2) * Hx_eff - M_zface(i, j, k, 0) * Hz_eff)
                 - M_zface(i, j, k, 1) * ( M_zface(i, j, k, 1) * Hz_eff - M_yface(i, j, k, 2) * Hy_eff));
-              });
+
+              // temporary normalized magnitude of M_zface field at the fixed point
+              // re-investigate the way we do Ms interp, in case we encounter the case where Ms changes across two adjacent cells that you are doing interp
+          amrex::Real mag_normalized = std::sqrt( std::pow(M_zface(i, j, k, 0),2.0_rt) + std::pow(M_zface(i, j, k, 1),2.0_rt) +
```
Comment (ajnonaka) [inline Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp]: indentation

**Review by ajnonaka — APPROVED** [review]
other than a few indentations (see comments) this looks as good as before!

**Review by jackieyao0114 — APPROVED** [review]
change `mag_normalized_error` to a user-input in the next PR.

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp [hunk 5]
```diff
@@ -125,6 +126,22 @@ void FiniteDifferenceSolver::MacroscopicEvolveM (
...
                 - M_xface(i, j, k, 1) * ( M_xface(i, j, k, 1) * Hz_eff - M_xface(i, j, k, 2) * Hy_eff));
 
+
+              // temporary normalized magnitude of M_xface field at the fixed point
+              // re-investigate the way we do Ms interp, in case we encounter the case where Ms changes across two adjacent cells that you are doing interp
+              amrex::Real mag_normalized = std::sqrt( std::pow(M_xface(i, j, k, 0),2.0) + std::pow(M_xface(i, j, k, 1),2.0) +
+                      std::pow(M_xface(i, j, k, 2),2.0) ) / MacroscopicProperties::macro_avg_to_face(i,j,k,amrex::IntVect(1,0,0),mag_Ms_arr);
+
+              // check the normalized error
+              if ( std::abs(1.-mag_normalized) > mag_normalized_error ){
```
Comment (RevathiJambunathan) [inline Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp]:
```suggestion
              if ( amrex::Math::abs(1._rt-mag_normalized) > mag_normalized_error ){
```

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp (hunk 1 above)
Comment (RevathiJambunathan) [inline Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp]:
```suggestion
    // Temporary value hard-coded for normalized error used for LLG.
    amrex::Real mag_normalized_error = 0.1_rt; // normalization error of M field for checking
```

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp [hunk 6]
```diff
@@ -170,6 +187,22 @@ void FiniteDifferenceSolver::MacroscopicEvolveM (
...
                 - M_yface(i, j, k, 1) * ( M_yface(i, j, k, 1) * Hz_eff - M_yface(i, j, k, 2) * Hy_eff));
 
+
+              // temporary normalized magnitude of M_yface field at the fixed point
+              // re-investigate the way we do Ms interp, in case we encounter the case where Ms changes across two adjacent cells that you are doing interp
+              amrex::Real mag_normalized = std::sqrt( std::pow(M_yface(i, j, k, 0),2.0) + std::pow(M_yface(i, j, k, 1),2.0) +
+                      std::pow(M_yface(i, j, k, 2),2.0) ) / MacroscopicProperties::macro_avg_to_face(i,j,k,amrex::IntVect(0,1,0),mag_Ms_arr);
+
+              // check the normalized error
+              if ( std::abs(1.-mag_normalized) > mag_normalized_error ){
```
Comment (RevathiJambunathan) [inline Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp]:
```suggestion
              if ( amrex::Math::abs(1._rt-mag_normalized) > mag_normalized_error ){
```

**Review by RevathiJambunathan — COMMENTED** [review]
@ruohai0925 Thanks for this PR! pending the few suggestions I ahve included, this PR is good to merge. You can "commit" the suggestions made here from the github GUI itself.
Let me know if you have any questions :)

**Review by RevathiJambunathan — APPROVED** [review]

**Pull request merged**